components/eam/src/physics/cam/cam_diagnostics.F90

```diff
@@ -2393,4 +2412,320 @@ subroutine diag_state_b4_phys_write (state)
+               a2(i) = -a2(i)*a1(i)**3
+               y(i) = q(i,mx(i)) - qstp(i,k)
+               tp(i,k) = tp(i,k) + a1(i)*y(i) + a2(i)*y(i)**2
+               call qsat_hPa(tp(i,k), p(i,k), estp(i), qstp(i,k))
+               tpv(i,k) =  tp(i,k) * (1._r8+1.608_r8*qstp(i,k)) / (1._r8+q(i,mx(i)))
+               buoy(i,k) = tpv(i,k) - tv(i,k) + tiedke_add
+               neg_buoy(i,k) = tv(i,k) - tpv(i,k) + tiedke_add
+            end if
+         end do
+      end do
+
+   ! main buoyancy calculation.
+      do k = pver - 1,1,-1
+         do i=1,ncol
+            if (k < lcl(i)) then
+               tv(i,k) = t(i,k)* (1._r8+1.608_r8*q(i,k))/ (1._r8+q(i,k))
+               qstp(i,k) = qstp(i,k+1)
+               tp(i,k) = tp(i,k+1)* (p(i,k)/p(i,k+1))**(0.2854_r8* (1._r8-0.28_r8*qstp(i,k)))
+               call qsat_hPa(tp(i,k), p(i,k), estp(i), qstp(i,k))
+               a1(i) = cpair/latvap + qstp(i,k)* (1._r8+qstp(i,k)/epsilo)*latvap*epsilo/ (rair*tp(i,k)**2)
+               a2(i) = .5_r8* (qstp(i,k)* (1._r8+2._r8/epsilo*qstp(i,k))* &
+                       (1._r8+qstp(i,k)/epsilo)*epsilo**2*latvap*latvap/ &
+                       (rair**2*tp(i,k)**4)-qstp(i,k)* &
+                       (1._r8+qstp(i,k)/epsilo)*2._r8*epsilo*latvap/ &
+                       (rair*tp(i,k)**3))
+               a1(i) = 1._r8/a1(i)
+               a2(i) = -a2(i)*a1(i)**3
+               y(i) = qstp(i,k+1) - qstp(i,k)
+               tp(i,k) = tp(i,k) + a1(i)*y(i) + a2(i)*y(i)**2
+               call qsat_hPa(tp(i,k), p(i,k), estp(i), qstp(i,k))
+               tpv(i,k) = tp(i,k) * (1._r8+1.608_r8*qstp(i,k))/(1._r8+q(i,mx(i)))
+               buoy(i,k) = tpv(i,k) - tv(i,k) + tiedke_add
+               neg_buoy(i,k) = tv(i,k) - tpv(i,k) + tiedke_add
+            end if
+         end do
+      end do
+
+      do k = 2,pver
+         do i = 1,ncol
+            if (k < lcl(i)) then
+               if (buoy(i,k+1) > 0._r8 .and. buoy(i,k) <= 0._r8) then
+                  knt(i) = min(num_cin,knt(i) + 1)
+                  lelten(i,knt(i)) = k
+               end if
+            end if
+         end do
+      end do
+
+   ! calculate convective available potential energy (cape).
+      do n = 1,num_cin
+         do k = 1,pver
+            do i = 1,ncol
+               if (k <= mx(i) .and. k > lelten(i,n)) then
+                  capeten(i,n) = capeten(i,n) + rair*buoy(i,k)*log(pf(i,k+1)/pf(i,k))
+               end if
+            end do
+         end do
+      end do
+
+   ! find maximum cape from all possible tentative capes from
+   ! one sounding,
+   ! and use it as the final cape, april 26, 1995
+      do n = 1,num_cin
+         do i = 1,ncol
+            if (capeten(i,n) > cape(i)) then
+               cape(i) = capeten(i,n)
+               lel(i) = lelten(i,n)
+            end if
+         end do
+      end do
+
+   ! Compute CIN based on information of levels computed above, which
+   !  is the buoyancy integrated from surface to the lauching level
```

Comment:
Going by this, shouldn't the launch level be mx rather than lcl in L2693? 

But taking a step back, I understood CIN to be the energy required to lift a parcel from the launch level to the level of free convection, rather than from the surface to LCL. But perhaps there's a different definition?